File: scala/leet_code/27.scala
/**
 * This is the Solution object. It contains a method to remove all instances of a specific value in-place from an array.
 * The method returns the new length of the array after the specified value has been removed.
 *
 * @example
 * val nums = Array(3,2,2,3)
 * val result = Solution.removeElement(nums, 3)
 * // result: 2
 * // nums: Array(2,2)
 *
 * @note The method modifies the input array by moving the non-specified elements to the front of the array.
 */
object Solution {
  /**
   * Removes all instances of a specific value in-place from an array.
   *
   * @param nums An array of integers.
   * @param `val` The value to be removed.
   * @return The new length of the array after the specified value has been removed.
   */
  def removeElement(nums: Array[Int], `val`: Int): Int = {
    var i = 0
    // Iterate over the array
    for (j <- 0 until nums.length) {
      // If the current element is not equal to the specified value
      if (nums(j) != `val`) {
        // Move the non-specified element to the front of the array
        nums(i) = nums(j)
        i += 1
      }
    }
    // Return the new length of the array
    i
  }
}

File: cpp/leet_code/27.cpp
/**
 * @file 27.cpp
 * @brief This file contains the solution for removing a specific element from an array.
 */
// Optimizing the code for speed and unrolling loops for efficiency.
#pragma GCC optimize("O3,unroll-loops")
#include <vector>

/**
 * @class Solution
 * @brief This class contains the method to remove a specific element from an array.
 */
class Solution {
public:
    /**
     * @brief This method removes all instances of a specific value in-place from an array and returns the new length.
     * @param nums A reference to the vector of integers.
     * @param val The value to be removed from the array.
     * @return The new length of the array after removing the specific value.
     * @note Do not allocate extra space for another array, you must do this by modifying the input array in-place with O(1) extra memory.
     * @note The order of elements can be changed. It doesn't matter what you leave beyond the new length.
     */
    static int removeElement(std::vector<int>& nums, int val) {
        int i = 0;
        // Iterate over the array
        for (int j = 0; j < nums.size(); j++) {
            // If the current element is not the value to be removed, move it to the beginning of the array
            if (nums[j] != val) {
                nums[i++] = nums[j];
            }
        }
        // Return the new length of the array
        return i;
    }
};
